llama_index/evaluation/benchmarks/beir.py

```diff
@@ -0,0 +1,90 @@
+from typing import Callable, List, Dict
+from llama_index.utils import get_cache_dir
+from llama_index.schema import Document
+from llama_index.indices.base_retriever import BaseRetriever
+import os
+import tqdm
+
+
+class BeirEvaluator:
+    """
+    Refer to: https://github.com/beir-cellar/beir for a full list of supported datasets
+    and a full description of BEIR.
+    """
+
+    def __init__(self) -> None:
+        try:
+            pass
+        except ImportError:
+            raise ImportError(
+                "Please install beir to use this feature: " "`pip install beir`",
+            )
+
+    def _download_datasets(self, datasets: List[str] = ["nfcorpus"]) -> Dict[str, str]:
+        from beir import util
+
+        cache_dir = get_cache_dir()
+
+        dataset_paths = {}
+        for dataset in datasets:
+            dataset_full_path = os.path.join(cache_dir, "datasets", "BeIR__" + dataset)
+            if not os.path.exists(dataset_full_path):
+                url = f""""https://public.ukp.informatik.tu-darmstadt.de\
+                        /thakur/BEIR/datasets/{dataset}.zip"""
+                util.download_and_unzip(url, dataset_full_path)
+            print("Dataset:", dataset, "downloaded at:", dataset_full_path)
+            dataset_paths[dataset] = os.path.join(dataset_full_path, dataset)
+        return dataset_paths
+
+    def run(
+        self,
+        create_retriever: Callable[[List[Document]], BaseRetriever],
+        datasets: List[str] = ["nfcorpus"],
+    ) -> None:
+        from beir.datasets.data_loader import GenericDataLoader
+        from beir.retrieval.evaluation import EvaluateRetrieval
+
+        dataset_paths = self._download_datasets(datasets)
+        for dataset in datasets:
+            dataset_path = dataset_paths[dataset]
+            print("Evaluating on dataset:", dataset)
+            print("-------------------------------------")
+
+            corpus, queries, qrels = GenericDataLoader(data_folder=dataset_path).load(
+                split="test"
+            )
+
+            documents = []
+            for id, val in corpus.items():
+                doc = Document(
+                    text=val["text"], metadata={"title": val["title"], "doc_id": id}
+                )
+                documents.append(doc)
+
+            retriever = create_retriever(documents)
+
+            print("Retriever created for: ", dataset)
+
+            print("Evaluating retriever on questions against qrels")
+
+            results = {}
+            for key, query in tqdm.tqdm(queries.items()):
+                nodes_with_score = retriever.retrieve(query)
+                results[key] = {
+                    node.node.metadata["doc_id"]: node.score
+                    for node in nodes_with_score
+                }
+
+            ndcg, map_, recall, precision = EvaluateRetrieval.evaluate(
+                qrels, results, [1, 10]
```

Comment:
Ok, last question before merging this (sorry lol)

Why 1 and 10 here? Shouldn't this be the top k of the retriever ?

Reply:
Well, technically, if one requests for top_k=100, based on the score, a different metric can be computed for k=1 to 100. I should explain this but it's a bit of advanced usage. I'll quickly add it in

Reply:
OK, I made it a parameter.